# <center> Diet Planning Problem

## <center> Group Members: 
<p> <center>Donnica Chick 19180223  <br>
Frances Li 93064061  <br>
Sara Wong 50313501   <br>
Hao Hsiang Yang 99407850   <br>
 

### <center> Table of Contents 
##### <center> 1. An Executive Summary
##### <center> 2. Introduction
##### <center> 3. Model formulations
##### <center> 4. Model results
##### <center> 5. Discussions/Recommendations/Conclusions


### <u>1. Executive Summary

The subject matter of this service is to provide solutions to a linear programming optimization problem. The problem is to come up with balanced and nutritious meals for children in the town of Starlight, while keeping the cost low in light of the town’s economic distress. Data that we used include the target daily nutrient range, the list of available food items and their costs as well as nutritional value. 

We utilized the Gurobi package in Python as our primary method of analysis. Initially, we found an optimal solution, however the meal plan was not balanced. Therefore, we added new constraints to ensure that the daily meal is more balanced using the Canada’s Food Guide. By leveraging the Gurobi package, modifying the constraints in the analysis, and using Python randomization, we constructed a 7-day meal plan that is optimal, nutritious, and contains a diversity of food items.

This report details the procedures performed and as well as the recommendation. Our deliverables include: (1) the optimal daily meal plan for one child at the lowest cost, (2) assessment on reasonableness of the optimal daily meal plan, and (3) a 7-day meal plan that includes variety. </p>

### <u>2. Introduction

<p> Sally Harper, a local school principal at the small town of Starlight, has hired our team as consultants to formulate a weekly meal plan for the local children. Principal Harper is leading the "Feeding Hope" initiative with the goal of providing nutritious meal plans for local children with minimal budget. Many families in Starlight have been struggling financially due to the closure of a local factory, which made it very difficult for parents to provide their children with sufficient amounts of meals. Harper's project has brought hope to the town by ensuring that children receive better care in their dietary intake. The project has attracted support from teachers, local farmers, and workers, who are volunteering their effort and time to assist with the project. However, monetary constraint is still the top concern, so it is essential to keep the cost low.

Harper has previously engaged a local university to build a linear programming optimization model that incorporated the various factors including the budget and necessary nutrition values. Based on the engagement agreement with Harper, our team will build a linear programming optimization model independently without having the model that Harper is currently using. The only data we received from Harper are the inputs related to the target daily nutrient ranges, the list of available food item options and their related costs and nutrient values. Harper will then perform a comparison between our model and her existing model. </p>

The following code blocks illustrate the step-by-step construction of a linear programming model in Python. First, a DataFrame is created to contain all food items and their nutrient values across different categories. Several lists are then generated from this DataFrame to automate the formulation of the objective function and constraints, allowing for an efficient construction of constraints and objective formulation for the model using a for loop.

In [2]:
#Data as taken from food_data.xlsx
data = {
    'Food': ['Broccoli', 'Carrots, Raw', 'Corn', 'Lettuce, Iceberg,Raw', 'Peppers, Sweet, Raw', 'Potatoes, Baked', 'Tofu', 'Roasted Chicken', 
             'Spaghetti W/ Sauce', 'Tomato,Red,Ripe,Raw', 'Apple, Raw, w/Skin', 'Banana', 'Grapes', 'Kiwifruit, Raw, Fresh', 'Oranges', 'Bagels', 
             'Wheat Bread', 'White Bread', '2% Lowfat Milk', 'Skim Milk', 'Poached Eggs', 'Scrambled Eggs', 'Turkey', 'Beef', 'Oatmeal', 'Couscous', 
             'White Rice', 'Macaroni, cooked', 'Pork', 'White Tuna in Water'],
    'Serving': ['10 Oz Pkg', '1/2 Cup Shredded', '1/2 Cup', '1 Leaf', '1 Pepper', '1/2 Cup', '1/4 block', '1 lb chicken', '1 1/2 Cup', 
                '1 Tomato, 2-3/5 In', '1 Fruit,3/Lb,Wo/Rf', '1 Fruit', '10 Grapes', '1 Medium', '1 Medium, 2-5/8 Diam', '1 Oz', '1 Slice', 
                '1 Slice', '1 Cup', '1 Cup', 'Lrg Egg', '1 Egg', '1 Oz', '1 Oz', '1 Cup', '1/2 Cup', '1/2 Cup', '1/2 Cup', '4 Oz', '3 Oz'], 
    'Calories (kcal)': [73.8, 23.7, 72.2, 2.6, 20.0, 171.5, 88.2, 277.4, 358.2, 25.8, 81.4, 104.9, 15.1, 46.4, 61.6, 78.0, 65.0, 65.0, 121.2, 85.5, 
                 74.5, 99.6, 56.4, 141.8, 145.1, 100.8, 103.0, 98.7, 710.8, 115.6], 
    'Fat (g)': [0.8, 0.1, 0.6, 0.0, 0.1, 0.2, 5.5, 10.8, 12.3, 0.4, 0.5, 0.5, 0.1, 0.3, 0.2, 0.5, 1.0, 1.0, 4.7, 0.4, 5.0, 7.3, 4.3, 12.8, 
                2.3, 0.1, 0.0, 0.5, 72.2, 2.1],
    'Sodium (mg)': [68.2, 19.2, 2.5, 1.8, 1.5, 15.2, 8.1, 125.6, 1237.1, 11.1, 0.0, 1.1, 0.5, 3.8, 0.0, 151.4, 134.5, 132.5, 121.8, 126.2, 
                    140.0, 168.0, 248.9, 461.7, 2.3, 4.5, 0.2, 0.7, 38.4, 333.2], 
    'Carbs (g)': [13.6, 5.6, 17.1, 0.4, 4.8, 39.9, 2.2, 0.0, 58.3, 5.7, 21.0, 26.7, 4.1, 11.3, 15.4, 15.1, 12.4, 11.8, 11.7, 11.9, 0.6, 1.3, 0.3, 0.8, 25.3, 20.9, 0.8, 19.8, 0.0, 0.0], 
    'Fiber (g)': [8.5, 1.6, 2.0, 0.3, 1.3, 3.2, 1.4, 0.0, 11.6, 1.4, 3.7, 2.7, 0.2, 2.6, 3.1, 0.6, 1.3, 1.1, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 4.0, 1.3, 22.3, 0.9, 0.0, 0.0], 
    'Protein (g)': [8.0, 0.6, 2.5, 0.2, 0.7, 3.7, 9.4, 42.2, 8.2, 1.0, 0.3, 1.2, 0.2, 0.8, 1.2, 3.0, 2.2, 2.3, 8.1, 8.4, 6.2, 6.7, 3.9, 5.4, 6.1, 3.4, 0.3, 3.3, 13.8, 22.7],
    'Vitamin A (IU)': [5867.4, 15471.0, 106.6, 66.0, 467.7, 0.0, 98.6, 77.4, 3055.2, 766.3, 73.1, 92.3, 24.0, 133.0, 268.6, 0.0, 0.0, 0.0, 500.2, 499.8, 316.0, 409.2, 0.0, 0.0, 37.4, 0.0, 2.1, 0.0, 14.7, 68.0], 
    'Vitamin C (mg)': [160.2, 5.1, 5.2, 0.8, 66.1, 15.6, 0.1, 0.0, 27.9, 23.5, 7.9, 10.4, 1.0, 74.5, 69.7, 0.0, 0.0, 0.0, 2.3, 2.4, 0.0, 0.1, 0.0, 10.8, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], 
    'Calcium (mg)': [159.0, 14.9, 3.3, 3.8, 6.7, 22.7, 121.8, 21.9, 80.2, 6.2, 9.7, 6.8, 3.4, 19.8, 52.4, 21.0, 10.8, 26.2, 296.7, 302.3, 24.5, 42.6, 23.8, 9.0, 18.7, 7.2, 0.0, 4.9, 59.9, 3.4], 
    'Iron (mg)': [2.3, 0.3, 0.3, 0.1, 0.3, 4.3, 6.2, 1.8, 2.3, 0.6, 0.2, 0.4, 0.1, 0.3, 0.1, 1.0, 0.7, 0.8, 0.1, 0.1, 0.7, 0.7, 0.4, 0.6, 1.6, 0.3, 7.9, 1.0, 0.4, 0.5], 
    'Cost ($/serving)': [0.16, 0.07, 0.18, 0.02, 0.53, 0.06, 0.31, 0.84, 0.78, 0.27, 0.24, 0.15, 0.32, 0.49, 0.15, 0.16, 0.05, 0.06, 0.23, 0.13, 0.08, 0.11, 0.15, 0.27, 0.82, 0.39, 0.08, 0.17, 0.81, 0.69]
}

In [3]:
import pandas as pd 
df = pd.DataFrame(data)
foods = df['Food'].tolist() #List of Food Names
columns = df.columns.tolist()[2:]  #Dietary requirements
column_name_gurobi = ['calories', 'fat', 'sodium', 'carbs', 'fiber', 'protein', 
                      'vitamin_A', 'vitamin_C', 'calcium', 'iron', 'cost'] #Clean list of food names for LP model 
col_servings = df['Serving'].tolist() #List of Servings

#List of all values in each column 
calories_df = df['Calories (kcal)'].tolist()
fat_df = df['Fat (g)'].tolist()
sodium_df = df['Sodium (mg)'].tolist()
carbs_df = df['Carbs (g)'].tolist()
fiber_df = df['Fiber (g)'].tolist()
protein_df = df['Protein (g)'].tolist()
vit_a_df = df['Vitamin A (IU)'].tolist()
vit_c_df = df['Vitamin C (mg)'].tolist()
calcium_df = df['Calcium (mg)'].tolist()
iron_df = df['Iron (mg)'].tolist()
cost_df = df['Cost ($/serving)'].tolist()


In [4]:
dietary_const = ['calories >= 1800', 'calories <= 2400', 'fat >= 60', 'fat <= 95', 'sodium >= 1200','sodium <= 2200', 'carbs >= 240', 'carbs <= 400',
                 'fibers <=35', 'fibers >= 30', 'protein >= 40', 'protein <= 55', 'vitamin_A>=2000', 'vitamin_A<=6000', 
                 'vitamin_C >= 45', 'vitamin_C <= 1200', 'calcium >= 1300', 'calcium <= 3000', 'iron >= 8', 'iron <= 40']

### <u>3. Model Formulations

##### <u>Question a
<p><u>Formulate (algebraically) an optimization problem to make a daily diet plan for each
child, with the goal of minimizing the total cost of the food while satisfying the general
nutritional requirements and additional considerations stated above.</p>

##### Building Optimization Model: 
##### Variables

In [5]:
print('Let')
for i in range(df.shape[0]): 
    var = 'a' + str(i) 
    print(f'{var} = {foods[i]}')

Let
a0 = Broccoli
a1 = Carrots, Raw
a2 = Corn
a3 = Lettuce, Iceberg,Raw
a4 = Peppers, Sweet, Raw
a5 = Potatoes, Baked
a6 = Tofu
a7 = Roasted Chicken
a8 = Spaghetti W/ Sauce
a9 = Tomato,Red,Ripe,Raw
a10 = Apple, Raw, w/Skin
a11 = Banana
a12 = Grapes
a13 = Kiwifruit, Raw, Fresh
a14 = Oranges
a15 = Bagels
a16 = Wheat Bread
a17 = White Bread
a18 = 2% Lowfat Milk
a19 = Skim Milk
a20 = Poached Eggs
a21 = Scrambled Eggs
a22 = Turkey
a23 = Beef
a24 = Oatmeal
a25 = Couscous
a26 = White Rice
a27 = Macaroni, cooked
a28 = Pork
a29 = White Tuna in Water


##### Objective and Constraints

In [6]:
print('Objective Function:\n')
lst = []
#Minimize cost equation
for i in range(df.shape[0]): 
     num = str(df.iloc[i]['Cost ($/serving)'])
     var = num + 'a' + str(i)
     lst.append(var)
     formula = " + ".join(lst)
print(f'Minimize ({formula}) \n')

print('Subject To: \n') #below are all our constraints

#Each nutritional content
for z in columns[:-1]: 
    lst = []
    for i in range(df.shape[0]):
        num = str(df.iloc[i][z])
        var = num + 'a' + str(i)
        lst.append(var)
    formula = " + ".join(lst)
    print(f'{column_name_gurobi[columns.index(z)]} = {formula} \n')

#Food item cannot be more than 30% of total calories
for i in range(df.shape[0]):   
    num = str(df.iloc[i]['Calories (kcal)'])
    value = "".join([num, 'a', str(i)])
    print(f'{value} <= 0.3*calories\n')

#Food item cannot be more than 30% of total protein
for i in range(df.shape[0]):
    num = str(df.iloc[i]['Protein (g)'])
    value = "".join([num, 'a', str(i)])
    print(f'{value} <= 0.3*protein\n')

for i in dietary_const: 
    print(i, '\n')


Objective Function:

Minimize (0.16a0 + 0.07a1 + 0.18a2 + 0.02a3 + 0.53a4 + 0.06a5 + 0.31a6 + 0.84a7 + 0.78a8 + 0.27a9 + 0.24a10 + 0.15a11 + 0.32a12 + 0.49a13 + 0.15a14 + 0.16a15 + 0.05a16 + 0.06a17 + 0.23a18 + 0.13a19 + 0.08a20 + 0.11a21 + 0.15a22 + 0.27a23 + 0.82a24 + 0.39a25 + 0.08a26 + 0.17a27 + 0.81a28 + 0.69a29) 

Subject To: 

calories = 73.8a0 + 23.7a1 + 72.2a2 + 2.6a3 + 20.0a4 + 171.5a5 + 88.2a6 + 277.4a7 + 358.2a8 + 25.8a9 + 81.4a10 + 104.9a11 + 15.1a12 + 46.4a13 + 61.6a14 + 78.0a15 + 65.0a16 + 65.0a17 + 121.2a18 + 85.5a19 + 74.5a20 + 99.6a21 + 56.4a22 + 141.8a23 + 145.1a24 + 100.8a25 + 103.0a26 + 98.7a27 + 710.8a28 + 115.6a29 

fat = 0.8a0 + 0.1a1 + 0.6a2 + 0.0a3 + 0.1a4 + 0.2a5 + 5.5a6 + 10.8a7 + 12.3a8 + 0.4a9 + 0.5a10 + 0.5a11 + 0.1a12 + 0.3a13 + 0.2a14 + 0.5a15 + 1.0a16 + 1.0a17 + 4.7a18 + 0.4a19 + 5.0a20 + 7.3a21 + 4.3a22 + 12.8a23 + 2.3a24 + 0.1a25 + 0.0a26 + 0.5a27 + 72.2a28 + 2.1a29 

sodium = 68.2a0 + 19.2a1 + 2.5a2 + 1.8a3 + 1.5a4 + 15.2a5 + 8.1a6 + 125.6a7 + 1237.

#### Optimization Model 

#### Base Model 
<p> There are no additional constraints besides those stated in the question.</p>

In the next two code blocks, we first define a results_df function. This function organizes the optimal solution generated by the model into a DataFrame, enhancing readability and presentation. In the subsequent code block, we implement the optimization model using Gurobi. The model operates according to the objective function and constraints previously defined, allowing us to find the optimal solution that satisfies all specified requirements.

In [7]:
def results_df(): 
    '''Function to put opitmized model results into a dataframe for better readability'''
    name_lst = []
    amt_lst = []
    unit_lst = []
    for i in range(n):
        if a[i].x != 0: 
            name_lst.append(foods[i])
            amt_lst.append(round(a[i].x, 2)) 
            unit_lst.append(col_servings[i])
    dic = {}
    dic['Food Item'] = name_lst
    dic['Servings'] = amt_lst
    dic['Serving Unit'] = unit_lst

    df_opt = pd.DataFrame(dic)
    return df_opt

In [8]:
import gurobipy as gp
from gurobipy import GRB
n=30
# Create a new model
model = gp.Model("Base Model")

# Coefficients for the objective function
cost_coeffs = [0.16, 0.07, 0.18, 0.02, 0.53, 0.06, 0.31, 0.84, 0.78, 0.27, 
               0.24, 0.15, 0.32, 0.49, 0.15, 0.16, 0.05, 0.06, 0.23, 0.13, 
               0.08, 0.11, 0.15, 0.27, 0.82, 0.39, 0.08, 0.17, 0.81, 0.69]

# Nutritional constraints coefficients
calories_coeffs = [73.8, 23.7, 72.2, 2.6, 20.0, 171.5, 88.2, 277.4, 358.2, 25.8, 
                   81.4, 104.9, 15.1, 46.4, 61.6, 78.0, 65.0, 65.0, 121.2, 85.5, 
                   74.5, 99.6, 56.4, 141.8, 145.1, 100.8, 103.0, 98.7, 710.8, 115.6]

fat_coeffs = [0.8, 0.1, 0.6, 0.0, 0.1, 0.2, 5.5, 10.8, 12.3, 0.4, 0.5, 0.5, 0.1, 
              0.3, 0.2, 0.5, 1.0, 1.0, 4.7, 0.4, 5.0, 7.3, 4.3, 12.8, 2.3, 0.1, 
              0.0, 0.5, 72.2, 2.1]

sodium_coeffs = [68.2, 19.2, 2.5, 1.8, 1.5, 15.2, 8.1, 125.6, 1237.1, 11.1, 0.0, 1.1, 
                 0.5, 3.8, 0.0, 151.4, 134.5, 132.5, 121.8, 126.2, 140.0, 168.0, 248.9, 
                 461.7, 2.3, 4.5, 0.2, 0.7, 38.4, 333.2]

carbs_coeffs = [13.6, 5.6, 17.1, 0.4, 4.8, 39.9, 2.2, 0.0, 58.3, 5.7, 21.0, 26.7, 
                4.1, 11.3, 15.4, 15.1, 12.4, 11.8, 11.7, 11.9, 0.6, 1.3, 0.3, 0.8, 
                25.3, 20.9, 0.8, 19.8, 0.0, 0.0]

fiber_coeffs = [8.5, 1.6, 2.0, 0.3, 1.3, 3.2, 1.4, 0.0, 11.6, 1.4, 3.7, 2.7, 0.2, 
                2.6, 3.1, 0.6, 1.3, 1.1, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 4.0, 1.3, 
                22.3, 0.9, 0.0, 0.0]

protein_coeffs = [8.0, 0.6, 2.5, 0.2, 0.7, 3.7, 9.4, 42.2, 8.2, 1.0, 0.3, 1.2, 0.2, 
                  0.8, 1.2, 3.0, 2.2, 2.3, 8.1, 8.4, 6.2, 6.7, 3.9, 5.4, 6.1, 3.4, 
                  0.3, 3.3, 13.8, 22.7]

vitamin_A_coeffs = [5867.4, 15471.0, 106.6, 66.0, 467.7, 0.0, 98.6, 77.4, 3055.2, 766.3, 
                    73.1, 92.3, 24.0, 133.0, 268.6, 0.0, 0.0, 0.0, 500.2, 499.8, 316.0, 
                    409.2, 0.0, 0.0, 37.4, 0.0, 2.1, 0.0, 14.7, 68.0]

vitamin_C_coeffs = [160.2, 5.1, 5.2, 0.8, 66.1, 15.6, 0.1, 0.0, 27.9, 23.5, 7.9, 10.4, 
                    1.0, 74.5, 69.7, 0.0, 0.0, 0.0, 2.3, 2.4, 0.0, 0.1, 0.0, 10.8, 
                    0.0, 0.0, 0.0, 0.0, 0.0, 0.0]

calcium_coeffs = [159.0, 14.9, 3.3, 3.8, 6.7, 22.7, 121.8, 21.9, 80.2, 6.2, 9.7, 6.8, 
                  3.4, 19.8, 52.4, 21.0, 10.8, 26.2, 296.7, 302.3, 24.5, 42.6, 23.8, 
                  9.0, 18.7, 7.2, 0.0, 4.9, 59.9, 3.4]

iron_coeffs = [2.3, 0.3, 0.3, 0.1, 0.3, 4.3, 6.2, 1.8, 2.3, 0.6, 0.2, 0.4, 0.1, 
               0.3, 0.1, 1.0, 0.7, 0.8, 0.1, 0.1, 0.7, 0.7, 0.4, 0.6, 1.6, 0.3, 
               7.9, 1.0, 0.4, 0.5]

# Decision variables
a = model.addVars(30, vtype=GRB.CONTINUOUS, name="a")

# Objective function: Minimize cost
model.setObjective(gp.quicksum(cost_coeffs[i] * a[i] for i in range(30)), GRB.MINIMIZE)

# Constraints

# Calories
model.addConstr(gp.quicksum(calories_coeffs[i] * a[i] for i in range(30)) >= 1800, "calories_lower")
model.addConstr(gp.quicksum(calories_coeffs[i] * a[i] for i in range(30)) <= 2400, "calories_upper")

# Fat
model.addConstr(gp.quicksum(fat_coeffs[i] * a[i] for i in range(30)) >= 60, "fat_lower")
model.addConstr(gp.quicksum(fat_coeffs[i] * a[i] for i in range(30)) <= 95, "fat_upper")

# Sodium
model.addConstr(gp.quicksum(sodium_coeffs[i] * a[i] for i in range(30)) >= 1200, "sodium_lower")
model.addConstr(gp.quicksum(sodium_coeffs[i] * a[i] for i in range(30)) <= 2200, "sodium_upper")

# Carbs
model.addConstr(gp.quicksum(carbs_coeffs[i] * a[i] for i in range(30)) >= 240, "carbs_lower")
model.addConstr(gp.quicksum(carbs_coeffs[i] * a[i] for i in range(30)) <= 400, "carbs_upper")

# Fiber
model.addConstr(gp.quicksum(fiber_coeffs[i] * a[i] for i in range(30)) >= 30, "fiber_lower")
model.addConstr(gp.quicksum(fiber_coeffs[i] * a[i] for i in range(30)) <= 35, "fiber_upper")

# Protein
model.addConstr(gp.quicksum(protein_coeffs[i] * a[i] for i in range(30)) >= 40, "protein_lower")
model.addConstr(gp.quicksum(protein_coeffs[i] * a[i] for i in range(30)) <= 55, "protein_upper")

# Vitamin A
model.addConstr(gp.quicksum(vitamin_A_coeffs[i] * a[i] for i in range(30)) >= 2000, "vitamin_A_lower")
model.addConstr(gp.quicksum(vitamin_A_coeffs[i] * a[i] for i in range(30)) <= 6000, "vitamin_A_upper")

# Vitamin C
model.addConstr(gp.quicksum(vitamin_C_coeffs[i] * a[i] for i in range(30)) >= 45, "vitamin_C_lower")
model.addConstr(gp.quicksum(vitamin_C_coeffs[i] * a[i] for i in range(30)) <= 1200, "vitamin_C_upper")

# Calcium
model.addConstr(gp.quicksum(calcium_coeffs[i] * a[i] for i in range(30)) >= 1300, "calcium_lower")
model.addConstr(gp.quicksum(calcium_coeffs[i] * a[i] for i in range(30)) <= 3000, "calcium_upper")

# Iron
model.addConstr(gp.quicksum(iron_coeffs[i] * a[i] for i in range(30)) >= 8, "iron_lower")
model.addConstr(gp.quicksum(iron_coeffs[i] * a[i] for i in range(30)) <= 40, "iron_upper")


# Constraints for each ingredient to be less than 30% total calories
total_calories = sum(calories_coeffs[i] * a[i] for i in range(30))
for i in range(30):
    model.addConstr((calories_coeffs[i] * a[i]) <= 0.3 * total_calories, name=f"calories_item_{i}")

# Constraints for each ingredient to be less than 30% total protein
total_protein = sum(protein_coeffs[i] * a[i] for i in range(30))
for i in range(30):
    model.addConstr((protein_coeffs[i] * a[i]) <= 0.3 * total_protein, name=f"protein_item_{i}")

# Optimize the model
model.setParam('OutputFlag', 0)
model.optimize()

# Print the solution
if model.status == GRB.OPTIMAL:
    print(f'\nMinimum cost required: ${model.objVal:.2f}')
    for i in range(30):
        print(f'{foods[i]}: {a[i].x}')
else:
    print("No optimal solution found.")

base = results_df()
base.index = base.index + 1
base_opt = model.objVal

Set parameter Username
Academic license - for non-commercial use only - expires 2025-09-02

Minimum cost required: $2.43
Broccoli: 0.0
Carrots, Raw: 0.0
Corn: 0.0
Lettuce, Iceberg,Raw: 0.0
Peppers, Sweet, Raw: 0.0
Potatoes, Baked: 3.14868804664723
Tofu: 0.0
Roasted Chicken: 0.0
Spaghetti W/ Sauce: 0.6109381088589423
Tomato,Red,Ripe,Raw: 0.0
Apple, Raw, w/Skin: 0.0
Banana: 0.09414308492014076
Grapes: 0.0
Kiwifruit, Raw, Fresh: 0.0
Oranges: 4.540457768798042
Bagels: 0.0
Wheat Bread: 0.0
White Bread: 0.0
2% Lowfat Milk: 2.0370370370370363
Skim Milk: 1.001073789921238
Poached Eggs: 0.0
Scrambled Eggs: 0.0
Turkey: 0.0
Beef: 0.0
Oatmeal: 0.0
Couscous: 0.0
White Rice: 0.15729644732766018
Macaroni, cooked: 0.0
Pork: 0.5668428942582175
White Tuna in Water: 0.0


### <u>4. Model results

##### <u>Question b
<p><u>Solve the formulation from part a) using Python/Gurobi to come up with a recommendation for the daily meal plan. Indicate the daily cost and the amount of each ingredient.</p>

In [9]:
print("Base Model Optimized Results")
print(f'Minimum cost required: ${round(base_opt,2)}')
base

Base Model Optimized Results
Minimum cost required: $2.43


Food Item  Servings          Serving Unit
1     Potatoes, Baked      3.15               1/2 Cup
2  Spaghetti W/ Sauce      0.61             1 1/2 Cup
3              Banana      0.09               1 Fruit
4             Oranges      4.54  1 Medium, 2-5/8 Diam
5      2% Lowfat Milk      2.04                 1 Cup
6           Skim Milk      1.00                 1 Cup
7          White Rice      0.16               1/2 Cup
8                Pork      0.57                  4 Oz

We built an optimization model above to find the optimal meal plan that meets Principal Harper’s main goal: to provide meals that offer all the necessary nutrients to children while minimizing cost. Additionally, to ensure a balanced diet, the optimal meal plan was designed so that no single food item makes up more than 30% of the total calories or protein intake.
After running our model, we found the optimal solution, which recommends the following meal components:

- 3.15 servings of Baked Potatoes
- 0.61 servings of Spaghetti with Sauce 
- 0.09 servings of Banana
- 4.54 servings of Oranges
- 2.04 servings of 2% Lowfat Milk
- 1.00 servings of Skim Milk
- 0.16 servings of White Rice
- 0.57 servings of Pork

This solution ensures that each child receives the necessary nutrients within the daily nutritional requirements, also at the lowest possible cost, which is $2.43.

### <u>5. Discussions/Recommendations/Conclusions

##### <u>Question c
<p><u>Discuss whether you think the recommendation from part b) sounds reasonable for a single day of food intake for a child.  If not, suggest some changes you would make to the model to make the recommendation for the day more realistic.  Implement your change and discuss the new solution.</p>

<p> However, as cost-efficient as this meal plan may be, it is not quite reasonable for a child to have 4.54 servings of oranges in a single day, and the overall meal plan does not seem balanced enough. Therefore, we adjusted our model by adding a few more constraints to make the recommended meal more realistic but also feasible for Principal Harper to implement the plan accordingly. 

Here’s our approach. First, we categorized all the food items into four categories according to Canada’s Food Guide, released by the Government of Canada[1]. These categories are Vegetables and Fruit; Grains; Milk and Alternatives; Meat and Alternatives.The main reason for introducing these categories to our solution was to ensure that the ingredients chosen to be used in the meal each day can be as balanced as possible. In the following step, we added constraints to limit the percentage of total servings for each category. Initially, we followed the recommendations for Canadian children aged 9 to 13 (the assumed age group under Principal Harper's care), which suggests 35% for Vegetables and Fruit; 35% for Grain; 20% for Milk and Alternatives; 10% for Meat and Alternatives. However, there turned out to be no optimal solution found after we ran the model using this approach.

To address this, we modified the categorization and adjusted the percentage constraints. Instead of four categories, we divided Vegetables and Fruit into separate groups, creating five categories in total: Grains, Milk and Alternatives, Meat and Alternatives, Vegetables, and Fruit. We also set a limit of no more than 30% of the total servings for any category. Additionally, we added a constraint that no single food item should exceed 3 servings, understanding that most children are unlikely to enjoy excessive amounts of the same food in one meal. </p>

In below codes, we implemented the food categorization by organizing items into four lists, each corresponding to a specific food category. For each category, we then identified the index of each item within the df DataFrame by iterating through these lists with a nested for loop.

In our updated Gurobi optimization model, we incorporated five additional constraints. These constraints impose proportional limits on each category to prevent any single category from dominating the solution, thereby ensuring a balanced distribution across all categories. With these new constraints in place, we derived an updated optimal solution that adheres to the newly defined category limits.

In [10]:
#Groups
group_meats = [ 'Tofu', 'Roasted Chicken','Turkey', 'Beef', 'Pork', 'White Tuna in Water','Poached Eggs', 'Scrambled Eggs']
group_dairy = ['2% Lowfat Milk', 'Skim Milk']
group_grains = ['Spaghetti W/ Sauce', 'Bagels', 'Wheat Bread', 'White Bread','Oatmeal', 'Couscous', 'White Rice', 'Macaroni, cooked']
group_vegs = ['Broccoli', 'Carrots, Raw', 'Corn', 'Lettuce, Iceberg,Raw', 'Peppers, Sweet, Raw', 'Potatoes, Baked', 
                     'Tomato,Red,Ripe,Raw']
group_fruits = ['Apple, Raw, w/Skin', 'Banana', 'Grapes','Kiwifruit, Raw, Fresh','Oranges']

group_lst = [group_meats ,group_dairy, group_grains, group_vegs, group_fruits]
group_lst_name = ['group_meats', 'group_dairy', 'group_grains', 'group_vegs', 'group_fruits']

#find the indexes of the foods in each group
for idx, group in enumerate(group_lst):
    lst = []
    for item in group: 
        lst.append(foods.index(item))
    name = "".join([str(group_lst_name[idx]), "_var"])
    print(f'{name} = {lst}') 


group_meats_var = [6, 7, 22, 23, 28, 29, 20, 21]
group_dairy_var = [18, 19]
group_grains_var = [8, 15, 16, 17, 24, 25, 26, 27]
group_vegs_var = [0, 1, 2, 3, 4, 5, 9]
group_fruits_var = [10, 11, 12, 13, 14]


<p>Output above copied and pasted below to turn into lists</p>

In [11]:
group_meats_var = [6, 7, 22, 23, 28, 29, 20, 21]
group_dairy_var = [18, 19]
group_grains_var = [8, 15, 16, 17, 24, 25, 26, 27]
group_vegs_var = [0, 1, 2, 3, 4, 5, 9]
group_fruits_var = [10, 11, 12, 13, 14]

In [12]:
# Create a new model
n = 30
model = gp.Model("nutrition")

# Coefficients for the objective function
cost_coeffs = [0.16, 0.07, 0.18, 0.02, 0.53, 0.06, 0.31, 0.84, 0.78, 0.27, 
               0.24, 0.15, 0.32, 0.49, 0.15, 0.16, 0.05, 0.06, 0.23, 0.13, 
               0.08, 0.11, 0.15, 0.27, 0.82, 0.39, 0.08, 0.17, 0.81, 0.69]

# Nutritional constraints coefficients
calories_coeffs = [73.8, 23.7, 72.2, 2.6, 20.0, 171.5, 88.2, 277.4, 358.2, 25.8, 
                   81.4, 104.9, 15.1, 46.4, 61.6, 78.0, 65.0, 65.0, 121.2, 85.5, 
                   74.5, 99.6, 56.4, 141.8, 145.1, 100.8, 103.0, 98.7, 710.8, 115.6]

fat_coeffs = [0.8, 0.1, 0.6, 0.0, 0.1, 0.2, 5.5, 10.8, 12.3, 0.4, 0.5, 0.5, 0.1, 
              0.3, 0.2, 0.5, 1.0, 1.0, 4.7, 0.4, 5.0, 7.3, 4.3, 12.8, 2.3, 0.1, 
              0.0, 0.5, 72.2, 2.1]

sodium_coeffs = [68.2, 19.2, 2.5, 1.8, 1.5, 15.2, 8.1, 125.6, 1237.1, 11.1, 0.0, 1.1, 
                 0.5, 3.8, 0.0, 151.4, 134.5, 132.5, 121.8, 126.2, 140.0, 168.0, 248.9, 
                 461.7, 2.3, 4.5, 0.2, 0.7, 38.4, 333.2]

carbs_coeffs = [13.6, 5.6, 17.1, 0.4, 4.8, 39.9, 2.2, 0.0, 58.3, 5.7, 21.0, 26.7, 
                4.1, 11.3, 15.4, 15.1, 12.4, 11.8, 11.7, 11.9, 0.6, 1.3, 0.3, 0.8, 
                25.3, 20.9, 0.8, 19.8, 0.0, 0.0]

fiber_coeffs = [8.5, 1.6, 2.0, 0.3, 1.3, 3.2, 1.4, 0.0, 11.6, 1.4, 3.7, 2.7, 0.2, 
                2.6, 3.1, 0.6, 1.3, 1.1, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 4.0, 1.3, 
                22.3, 0.9, 0.0, 0.0]

protein_coeffs = [8.0, 0.6, 2.5, 0.2, 0.7, 3.7, 9.4, 42.2, 8.2, 1.0, 0.3, 1.2, 0.2, 
                  0.8, 1.2, 3.0, 2.2, 2.3, 8.1, 8.4, 6.2, 6.7, 3.9, 5.4, 6.1, 3.4, 
                  0.3, 3.3, 13.8, 22.7]

vitamin_A_coeffs = [5867.4, 15471.0, 106.6, 66.0, 467.7, 0.0, 98.6, 77.4, 3055.2, 766.3, 
                    73.1, 92.3, 24.0, 133.0, 268.6, 0.0, 0.0, 0.0, 500.2, 499.8, 316.0, 
                    409.2, 0.0, 0.0, 37.4, 0.0, 2.1, 0.0, 14.7, 68.0]

vitamin_C_coeffs = [160.2, 5.1, 5.2, 0.8, 66.1, 15.6, 0.1, 0.0, 27.9, 23.5, 7.9, 10.4, 
                    1.0, 74.5, 69.7, 0.0, 0.0, 0.0, 2.3, 2.4, 0.0, 0.1, 0.0, 10.8, 
                    0.0, 0.0, 0.0, 0.0, 0.0, 0.0]

calcium_coeffs = [159.0, 14.9, 3.3, 3.8, 6.7, 22.7, 121.8, 21.9, 80.2, 6.2, 9.7, 6.8, 
                  3.4, 19.8, 52.4, 21.0, 10.8, 26.2, 296.7, 302.3, 24.5, 42.6, 23.8, 
                  9.0, 18.7, 7.2, 0.0, 4.9, 59.9, 3.4]

iron_coeffs = [2.3, 0.3, 0.3, 0.1, 0.3, 4.3, 6.2, 1.8, 2.3, 0.6, 0.2, 0.4, 0.1, 
               0.3, 0.1, 1.0, 0.7, 0.8, 0.1, 0.1, 0.7, 0.7, 0.4, 0.6, 1.6, 0.3, 
               7.9, 1.0, 0.4, 0.5]

# Decision variables
a = model.addVars(30, vtype=GRB.CONTINUOUS, name="a")

# Objective function: Minimize cost
model.setObjective(gp.quicksum(cost_coeffs[i] * a[i] for i in range(30)), GRB.MINIMIZE)

# Constraints

# Calories
model.addConstr(gp.quicksum(calories_coeffs[i] * a[i] for i in range(30)) >= 1800, "calories_lower")
model.addConstr(gp.quicksum(calories_coeffs[i] * a[i] for i in range(30)) <= 2400, "calories_upper")

# Fat
model.addConstr(gp.quicksum(fat_coeffs[i] * a[i] for i in range(30)) >= 60, "fat_lower")
model.addConstr(gp.quicksum(fat_coeffs[i] * a[i] for i in range(30)) <= 95, "fat_upper")

# Sodium
model.addConstr(gp.quicksum(sodium_coeffs[i] * a[i] for i in range(30)) >= 1200, "sodium_lower")
model.addConstr(gp.quicksum(sodium_coeffs[i] * a[i] for i in range(30)) <= 2200, "sodium_upper")

# Carbs
model.addConstr(gp.quicksum(carbs_coeffs[i] * a[i] for i in range(30)) >= 240, "carbs_lower")
model.addConstr(gp.quicksum(carbs_coeffs[i] * a[i] for i in range(30)) <= 400, "carbs_upper")

# Fiber
model.addConstr(gp.quicksum(fiber_coeffs[i] * a[i] for i in range(30)) >= 30, "fiber_lower")
model.addConstr(gp.quicksum(fiber_coeffs[i] * a[i] for i in range(30)) <= 35, "fiber_upper")

# Protein
model.addConstr(gp.quicksum(protein_coeffs[i] * a[i] for i in range(30)) >= 40, "protein_lower")
model.addConstr(gp.quicksum(protein_coeffs[i] * a[i] for i in range(30)) <= 55, "protein_upper")

# Vitamin A
model.addConstr(gp.quicksum(vitamin_A_coeffs[i] * a[i] for i in range(30)) >= 2000, "vitamin_A_lower")
model.addConstr(gp.quicksum(vitamin_A_coeffs[i] * a[i] for i in range(30)) <= 6000, "vitamin_A_upper")

# Vitamin C
model.addConstr(gp.quicksum(vitamin_C_coeffs[i] * a[i] for i in range(30)) >= 45, "vitamin_C_lower")
model.addConstr(gp.quicksum(vitamin_C_coeffs[i] * a[i] for i in range(30)) <= 1200, "vitamin_C_upper")

# Calcium
model.addConstr(gp.quicksum(calcium_coeffs[i] * a[i] for i in range(30)) >= 1300, "calcium_lower")
model.addConstr(gp.quicksum(calcium_coeffs[i] * a[i] for i in range(30)) <= 3000, "calcium_upper")

# Iron
model.addConstr(gp.quicksum(iron_coeffs[i] * a[i] for i in range(30)) >= 8, "iron_lower")
model.addConstr(gp.quicksum(iron_coeffs[i] * a[i] for i in range(30)) <= 40, "iron_upper")

total_calories = sum(calories_coeffs[i] * a[i] for i in range(30))

# Constraint for each item to be less than 30% of total calories
for i in range(30):
    model.addConstr((calories_coeffs[i] * a[i]) <= 0.3 * total_calories, name=f"calories_item_{i}")

total_protein = sum(protein_coeffs[i] * a[i] for i in range(30))

# Constraint for each item to be less than 30% of total protein
for i in range(30):
    model.addConstr((protein_coeffs[i] * a[i]) <= 0.3 * total_protein, name=f"protein_item_{i}")

# Constraint for each item to be less than 3 servings
for i in range(30):
    model.addConstr(a[i] <= 3, name=f"item_constraint{i}")

# Constraint for each group to be less than 30% of the total servings
total_groups = gp.quicksum(a[i] for i in range(30))

model.addConstr(gp.quicksum(a[group_meats_var[i]] for i in range(len(group_meats_var))) <= 0.3 * total_groups)
model.addConstr(gp.quicksum(a[group_grains_var[i]] for i in range(len(group_grains_var))) <= 0.3 * total_groups)
model.addConstr(gp.quicksum(a[group_dairy_var[i]] for i in range(len(group_dairy_var))) <= 0.3 * total_groups)
model.addConstr(gp.quicksum(a[group_vegs_var[i]] for i in range(len(group_vegs_var))) <= 0.3 * total_groups)
model.addConstr(gp.quicksum(a[group_fruits_var[i]] for i in range(len(group_fruits_var))) <= 0.3 * total_groups)


# Optimize the model
model.setParam('OutputFlag', 0)
model.optimize()

# Print the solution
if model.status == GRB.OPTIMAL:
    print(f'\nMinimum cost required: ${model.objVal:.2f}')
    for i in range(30):
        print(f'{foods[i]}: {a[i].x}')
else:
    print("No optimal solution found.")
model_c = results_df()
model_c.index = model_c.index + 1
model_c_opt = model.objVal


Minimum cost required: $2.60
Broccoli: 0.0
Carrots, Raw: 0.13059432198177207
Corn: 0.0
Lettuce, Iceberg,Raw: 0.6752817545202033
Peppers, Sweet, Raw: 0.0
Potatoes, Baked: 2.9245166384011676
Tofu: 0.0
Roasted Chicken: 0.0
Spaghetti W/ Sauce: 0.5525438933732834
Tomato,Red,Ripe,Raw: 0.0
Apple, Raw, w/Skin: 2.908841916931123
Banana: 0.0
Grapes: 0.0
Kiwifruit, Raw, Fresh: 0.0
Oranges: 0.8215507979720198
Bagels: 0.0
Wheat Bread: 0.0
White Bread: 0.0
2% Lowfat Milk: 2.0370370370370376
Skim Milk: 1.5722710799719268
Poached Eggs: 0.0
Scrambled Eggs: 0.0
Turkey: 0.0
Beef: 0.0
Oatmeal: 0.0
Couscous: 0.0
White Rice: 0.24712912942156345
Macaroni, cooked: 0.0
Pork: 0.5648758134003794
White Tuna in Water: 0.0


In [13]:
print(f'Minimum cost required: ${round(model_c_opt,2)}')
model_c

Minimum cost required: $2.6


Food Item  Servings          Serving Unit
1           Carrots, Raw      0.13      1/2 Cup Shredded
2   Lettuce, Iceberg,Raw      0.68                1 Leaf
3        Potatoes, Baked      2.92               1/2 Cup
4     Spaghetti W/ Sauce      0.55             1 1/2 Cup
5     Apple, Raw, w/Skin      2.91    1 Fruit,3/Lb,Wo/Rf
6                Oranges      0.82  1 Medium, 2-5/8 Diam
7         2% Lowfat Milk      2.04                 1 Cup
8              Skim Milk      1.57                 1 Cup
9             White Rice      0.25               1/2 Cup
10                  Pork      0.56                  4 Oz

After running our adjusted model, we arrived at a new optimal solution, which recommends the following meal components:
- 0.13 servings of Carrots, Raw
- 0.68 servings of Lettuce, Iceberg,Raw
- 2.92 servings of Potatoes, Baked
- 0.55 servings of Spaghetti W/ Sauce
- 2.91 servings of Apple, Raw, w/Skin
- 0.82 servings of Oranges
- 2.04 servings of 2% Lowfat Milk
- 1.57 servings of Skim Milk
- 0.25 servings of White Rice
- 0.56 servings of Pork

In this updated plan, food items from all categories are included, and no item exceeds 3 servings. While this new approach does not exactly follow the Canada’s Food Guide, it provides a more balanced and diverse meal plan compared to the solution from Part B, and it only increases the cost by $0.17. We believe this new model offers a more practical and favorable solution overall.


### Discussion and Analysis:

##### <u>Question d
<p><u>Whereas part c) we consider whether the meal plan for the one day is sensible or not and how you might get to a reasonable plan for even one day.  However, Principal Harper also knows that kids don’t like to eat the same exact thing every day, so for part d) she asks us to come up with a reasonable dietary plan for each day of an entire week.  Present a 7-day dietary plan, and indicate any model changes that went into constructing each day’s plan.</p>

Besides the one-day meal plan, Principal Harper also requested our team to devise a reasonable meal plan for an entire week. The important consideration here is to ensure there is variety in the food items, as kids do not like to eat the same thing every day.
 
To obtain variety of dietary meeal plans over 7 days, we utilized the random module in Python to generate a randomized food list. By setting the seeds, we can ensure we generate consistent randomized list every run. The randomization selected 80% from each dietary group to ensure a balanced diet and then returns a combined list of all the food items. Because we are already selecting food items from each dietary group, we dropped the constraints on limiting 30% on each dietary group. We kept the constraint of limiting 3 servings per food item to avoid over-consumption of any single food item. All other constraints for the nutrient intake stay the same as in part B. 

In [132]:
import random
def random_food_lst(str): 
    '''Randomly selects 80% of food items from each group and returns a combined list'''
    total_lst = []
    for idx, g in enumerate(group_lst): 
        lst = random.sample(g, k=round(0.8*len(g)))
        #print(f'{str}_{group_lst_name[idx]} = {lst}')
        total_lst.append(lst)
    combined_lst = [item for sublst in total_lst for item in sublst]
    #print(combined_lst)
    return combined_lst

In [133]:
for z in range(15): 
    random.seed(z)
    rand_food_lst = random_food_lst('new_day')
    lst_idx = []
    for i in rand_food_lst: 
        lst_idx.append(foods.index(i))
        lst_idx.sort()

    rand_food_lst = []
    for i in lst_idx: 
        rand_food_lst.append(foods[i])
    rand_food_lst

    
    n = len(lst_idx)
    # Create a new model
    model = gp.Model("Base Model")

    # Coefficients for the objective function
    cost_coeff = [cost_df[i] for i in lst_idx]
    
    # Nutritional constraints coefficients
    calories_coeffs = [calories_df[i] for i in lst_idx]
    fat_coeffs = [fat_df[i] for i in lst_idx]
    sodium_coeffs = [sodium_df[i] for i in lst_idx]
    carbs_coeffs = [carbs_df[i] for i in lst_idx]
    fiber_coeffs = [fiber_df[i] for i in lst_idx]
    protein_coeffs = [protein_df[i] for i in lst_idx]
    vitamin_A_coeffs = [vit_a_df[i] for i in lst_idx]
    vitamin_C_coeffs = [vit_c_df[i] for i in lst_idx]
    calcium_coeffs = [calcium_df[i] for i in lst_idx]
    iron_coeffs = [iron_df[i] for i in lst_idx]

    # Decision variables
    a = model.addVars(n, vtype=GRB.CONTINUOUS, name="a")
    
    # Objective function: Minimize cost
    model.setObjective(gp.quicksum(cost_coeff[i] * a[i] for i in range(n)), GRB.MINIMIZE)
    
    # Constraints
    
    # Calories
    model.addConstr(gp.quicksum(calories_coeffs[i] * a[i] for i in range(n)) >= 1800, "calories_lower")
    model.addConstr(gp.quicksum(calories_coeffs[i] * a[i] for i in range(n)) <= 2400, "calories_upper")
    
    # Fat
    model.addConstr(gp.quicksum(fat_coeffs[i] * a[i] for i in range(n)) >= 60, "fat_lower")
    model.addConstr(gp.quicksum(fat_coeffs[i] * a[i] for i in range(n)) <= 95, "fat_upper")
    
    # Sodium
    model.addConstr(gp.quicksum(sodium_coeffs[i] * a[i] for i in range(n)) >= 1200, "sodium_lower")
    model.addConstr(gp.quicksum(sodium_coeffs[i] * a[i] for i in range(n)) <= 2200, "sodium_upper")
    
    # Carbs
    model.addConstr(gp.quicksum(carbs_coeffs[i] * a[i] for i in range(n)) >= 240, "carbs_lower")
    model.addConstr(gp.quicksum(carbs_coeffs[i] * a[i] for i in range(n)) <= 400, "carbs_upper")
    
    # Fiber
    model.addConstr(gp.quicksum(fiber_coeffs[i] * a[i] for i in range(n)) >= 30, "fiber_lower")
    model.addConstr(gp.quicksum(fiber_coeffs[i] * a[i] for i in range(n)) <= 35, "fiber_upper")
    
    # Protein
    model.addConstr(gp.quicksum(protein_coeffs[i] * a[i] for i in range(n)) >= 40, "protein_lower")
    model.addConstr(gp.quicksum(protein_coeffs[i] * a[i] for i in range(n)) <= 55, "protein_upper")
    
    # Vitamin A
    model.addConstr(gp.quicksum(vitamin_A_coeffs[i] * a[i] for i in range(n)) >= 2000, "vitamin_A_lower")
    model.addConstr(gp.quicksum(vitamin_A_coeffs[i] * a[i] for i in range(n)) <= 6000, "vitamin_A_upper")
    
    # Vitamin C
    model.addConstr(gp.quicksum(vitamin_C_coeffs[i] * a[i] for i in range(n)) >= 45, "vitamin_C_lower")
    model.addConstr(gp.quicksum(vitamin_C_coeffs[i] * a[i] for i in range(n)) <= 1200, "vitamin_C_upper")
    
    # Calcium
    model.addConstr(gp.quicksum(calcium_coeffs[i] * a[i] for i in range(n)) >= 1300, "calcium_lower")
    model.addConstr(gp.quicksum(calcium_coeffs[i] * a[i] for i in range(n)) <= 3000, "calcium_upper")
    
    # Iron
    model.addConstr(gp.quicksum(iron_coeffs[i] * a[i] for i in range(n)) >= 8, "iron_lower")
    model.addConstr(gp.quicksum(iron_coeffs[i] * a[i] for i in range(n)) <= 40, "iron_upper")
    
    
    # Constraints for each ingredient to be less than 30% total calories
    total_calories = sum(calories_coeffs[i] * a[i] for i in range(n))
    for i in range(n):
        model.addConstr((calories_coeffs[i] * a[i]) <= 0.3 * total_calories, name=f"calories_item_{i}")
    
    # Constraints for each ingredient to be less than 30% total protein
    total_protein = sum(protein_coeffs[i] * a[i] for i in range(n))
    for i in range(n):
        model.addConstr((protein_coeffs[i] * a[i]) <= 0.3 * total_protein, name=f"protein_item_{i}")
    
    
    for i in range(n):
        model.addConstr(a[i] <= 3 , name=f"item_constraint{i}")
    
    # Optimize the model
    model.setParam('OutputFlag', 0)
    model.optimize()
    
    # Print the solution
    if model.status != GRB.OPTIMAL:
        continue
    else: 
        #for i in range(n):
            #print(f'{rand_food_lst[i]}: {a[i].x}')
        print(f'\nMinimum cost required: ${model.objVal:.2f}')
        
    model_d_opt = model.objVal
    print(f'Seed: {z}')
    model_d = results_df()
    model_d.index = model_d.index + 1
    print(model_d)


Minimum cost required: $2.45
Seed: 1
             Food Item  Servings        Serving Unit
1  Peppers, Sweet, Raw      2.50            1 Pepper
2      Roasted Chicken      0.59        1 lb chicken
3   Apple, Raw, w/Skin      1.83  1 Fruit,3/Lb,Wo/Rf
4               Grapes      3.00           10 Grapes
5               Bagels      2.04                1 Oz
6          Wheat Bread      1.28             1 Slice
7         Poached Eggs      0.27             Lrg Egg
8               Turkey      0.56                1 Oz

Minimum cost required: $2.50
Seed: 2
             Food Item  Servings        Serving Unit
1  Peppers, Sweet, Raw      3.00            1 Pepper
2      Roasted Chicken      0.59        1 lb chicken
3  Tomato,Red,Ripe,Raw      1.20  1 Tomato, 2-3/5 In
4               Grapes      3.00           10 Grapes
5               Bagels      2.04                1 Oz
6          Wheat Bread      1.25             1 Slice
7       Scrambled Eggs      0.22               1 Egg
8                 Beef 

We generated 12 optimal solutions from 15 randomized food lists (as 3 of the lists did not yield optimal solutions). We then compared the diversity of food items across these 12 optimal solutions and selected the top 6 solutions with the highest diversity.

Below codes were used to conduct a similarity comparison analysis to measure the similarity between each pair. Using this similarity metric, we then select the six most diverse results to ensure a range of distinct solutions. Following this selection, we rerun the optimization model for each of these six diverse results to generate a comprehensive list of food items corresponding to each solution.

steven: again, very interesting approach.  it would help to explain more in words the idea behind the similarity metric. 

In [136]:
# List of 12 food lists
food_lists = [
    ['Apple, Raw, w/Skin', 'Bagels', 'Grapes', 'Peppers, Sweet, Raw', 'Poached Eggs', 'Roasted Chicken', 'Turkey', 'Wheat Bread'],
    ['Bagels', 'Beef', 'Grapes', 'Peppers, Sweet, Raw', 'Roasted Chicken', 'Scrambled Eggs', 'Tomato,Red,Ripe,Raw', 'Wheat Bread'],
    ['Bagels', 'Banana', 'Beef', 'Peppers, Sweet, Raw', 'Scrambled Eggs', 'Spaghetti W/ Sauce', 'Tofu', 'Wheat Bread'],
    ['Bagels', 'Oranges', 'Peppers, Sweet, Raw', 'Poached Eggs', 'Spaghetti W/ Sauce', 'Tofu', 'Tomato,Red,Ripe,Raw', 'Turkey'],
    ['Apple, Raw, w/Skin', 'Broccoli', 'Carrots, Raw', 'Peppers, Sweet, Raw', 'Roasted Chicken', 'Tomato,Red,Ripe,Raw', 'Turkey', 'Wheat Bread', 'White Bread'],
    ['Apple, Raw, w/Skin', 'Bagels', 'Beef', 'Grapes', 'Peppers, Sweet, Raw', 'Roasted Chicken', 'Scrambled Eggs', 'Wheat Bread'],
    ['Bagels', 'Grapes', 'Oranges', 'Poached Eggs', 'Potatoes, Baked', 'Spaghetti W/ Sauce', 'Tomato,Red,Ripe,Raw', 'Turkey'],
    ['Apple, Raw, w/Skin', 'Bagels', 'Beef', 'Grapes', 'Kiwifruit, Raw, Fresh', 'Oranges', 'Potatoes, Baked', 'Roasted Chicken', 'Scrambled Eggs', 'Skim Milk', 'Spaghetti W/ Sauce'],
    ['Apple, Raw, w/Skin', 'Bagels', 'Carrots, Raw', 'Kiwifruit, Raw, Fresh', 'Oranges', 'Poached Eggs', 'Roasted Chicken', 'Tomato,Red,Ripe,Raw', 'Turkey'],
    ['Bagels', 'Banana', 'Oranges', 'Peppers, Sweet, Raw', 'Scrambled Eggs', 'Spaghetti W/ Sauce', 'Tofu', 'Turkey'],
    ['Grapes', 'Kiwifruit, Raw, Fresh', 'Peppers, Sweet, Raw', 'Poached Eggs', 'Potatoes, Baked', 'Roasted Chicken', 'Spaghetti W/ Sauce', 'Turkey'],
    ['Bagels', 'Oranges', 'Peppers, Sweet, Raw', 'Poached Eggs', 'Spaghetti W/ Sauce', 'Tofu', 'Tomato,Red,Ripe,Raw', 'Turkey']
]

# Function to calculate overlap between two lists
def calculate_overlap(list1, list2):
    return len(set(list1) & set(list2))

# Function to select 6 lists with the most diversity and return their original indices
def select_most_diverse_lists(food_lists, num_to_select=6):
    # Create a list of tuples (index, food_list)
    indexed_food_lists = list(enumerate(food_lists))
    
    selected_lists = []
    selected_indices = []
    remaining_lists = indexed_food_lists.copy()

    # Start by selecting the first list (arbitrary choice)
    first_index, first_list = remaining_lists.pop(0)
    selected_lists.append(first_list)
    selected_indices.append(first_index)

    # Iteratively select lists that minimize overlap with the selected ones
    for _ in range(num_to_select - 1):
        # Find the list with the least overlap with already selected lists
        min_overlap_list = None
        min_overlap_value = float('inf')
        min_overlap_index = None
        
        for index, candidate_list in remaining_lists:
            # Calculate total overlap with all previously selected lists
            total_overlap = sum(calculate_overlap(candidate_list, selected_list) for selected_list in selected_lists)
            
            # Keep track of the list with the least overlap
            if total_overlap < min_overlap_value:
                min_overlap_value = total_overlap
                min_overlap_list = candidate_list
                min_overlap_index = index
        
        # Add the list with the least overlap to the selected lists
        selected_lists.append(min_overlap_list)
        selected_indices.append(min_overlap_index)

        # Remove the selected list from remaining lists
        remaining_lists = [item for item in remaining_lists if item[0] != min_overlap_index]

    return selected_lists, selected_indices

# Function to sequence the lists to minimize overlap between consecutive days
def sequence_lists(selected_lists):
    sequenced_lists = []
    remaining_lists = selected_lists.copy()

    # Start by choosing the first list arbitrarily
    sequenced_lists.append(remaining_lists.pop(0))

    # Greedily select the next list that has the least overlap with the last selected list
    while remaining_lists:
        last_list = sequenced_lists[-1]
        min_overlap_list = None
        min_overlap_value = float('inf')

        for candidate_list in remaining_lists:
            # Calculate overlap with the last selected list
            overlap = calculate_overlap(last_list, candidate_list)
            if overlap < min_overlap_value:
                min_overlap_value = overlap
                min_overlap_list = candidate_list
        
        # Add the list with the least overlap to the sequence
        sequenced_lists.append(min_overlap_list)
        remaining_lists.remove(min_overlap_list)

    return sequenced_lists

# Get the 6 most diverse lists and their original indices
most_diverse_lists, most_diverse_indices = select_most_diverse_lists(food_lists, num_to_select=6)

# Sequence the lists to minimize overlap between consecutive days
sequenced_meal_plans = sequence_lists(most_diverse_lists)

# Output the sequenced lists with their original indices
for i, (index, meal_plan) in enumerate(zip(most_diverse_indices, most_diverse_lists), start=1):
    print(f"Day {i+1} Meal Plan (Seed {index+1})")

Day 2 Meal Plan (Seed 1)
Day 3 Meal Plan (Seed 3)
Day 4 Meal Plan (Seed 7)
Day 5 Meal Plan (Seed 5)
Day 6 Meal Plan (Seed 8)
Day 7 Meal Plan (Seed 10)


In [137]:
#rerun to get just the dietary plans for chosen seeds
for z in [1,3,7,5,8,10]: 
    random.seed(z)
    rand_food_lst = random_food_lst('new_day')
    lst_idx = []
    for i in rand_food_lst: 
        lst_idx.append(foods.index(i))
        lst_idx.sort()

    rand_food_lst = []
    for i in lst_idx: 
        rand_food_lst.append(foods[i])
    rand_food_lst

    
    n = len(lst_idx)
    # Create a new model
    model = gp.Model("Base Model")

    # Coefficients for the objective function
    cost_coeff = [cost_df[i] for i in lst_idx]
    
    # Nutritional constraints coefficients
    calories_coeffs = [calories_df[i] for i in lst_idx]
    fat_coeffs = [fat_df[i] for i in lst_idx]
    sodium_coeffs = [sodium_df[i] for i in lst_idx]
    carbs_coeffs = [carbs_df[i] for i in lst_idx]
    fiber_coeffs = [fiber_df[i] for i in lst_idx]
    protein_coeffs = [protein_df[i] for i in lst_idx]
    vitamin_A_coeffs = [vit_a_df[i] for i in lst_idx]
    vitamin_C_coeffs = [vit_c_df[i] for i in lst_idx]
    calcium_coeffs = [calcium_df[i] for i in lst_idx]
    iron_coeffs = [iron_df[i] for i in lst_idx]

    # Decision variables
    a = model.addVars(n, vtype=GRB.CONTINUOUS, name="a")
    
    # Objective function: Minimize cost
    model.setObjective(gp.quicksum(cost_coeff[i] * a[i] for i in range(n)), GRB.MINIMIZE)
    
    # Constraints
    
    # Calories
    model.addConstr(gp.quicksum(calories_coeffs[i] * a[i] for i in range(n)) >= 1800, "calories_lower")
    model.addConstr(gp.quicksum(calories_coeffs[i] * a[i] for i in range(n)) <= 2400, "calories_upper")
    
    # Fat
    model.addConstr(gp.quicksum(fat_coeffs[i] * a[i] for i in range(n)) >= 60, "fat_lower")
    model.addConstr(gp.quicksum(fat_coeffs[i] * a[i] for i in range(n)) <= 95, "fat_upper")
    
    # Sodium
    model.addConstr(gp.quicksum(sodium_coeffs[i] * a[i] for i in range(n)) >= 1200, "sodium_lower")
    model.addConstr(gp.quicksum(sodium_coeffs[i] * a[i] for i in range(n)) <= 2200, "sodium_upper")
    
    # Carbs
    model.addConstr(gp.quicksum(carbs_coeffs[i] * a[i] for i in range(n)) >= 240, "carbs_lower")
    model.addConstr(gp.quicksum(carbs_coeffs[i] * a[i] for i in range(n)) <= 400, "carbs_upper")
    
    # Fiber
    model.addConstr(gp.quicksum(fiber_coeffs[i] * a[i] for i in range(n)) >= 30, "fiber_lower")
    model.addConstr(gp.quicksum(fiber_coeffs[i] * a[i] for i in range(n)) <= 35, "fiber_upper")
    
    # Protein
    model.addConstr(gp.quicksum(protein_coeffs[i] * a[i] for i in range(n)) >= 40, "protein_lower")
    model.addConstr(gp.quicksum(protein_coeffs[i] * a[i] for i in range(n)) <= 55, "protein_upper")
    
    # Vitamin A
    model.addConstr(gp.quicksum(vitamin_A_coeffs[i] * a[i] for i in range(n)) >= 2000, "vitamin_A_lower")
    model.addConstr(gp.quicksum(vitamin_A_coeffs[i] * a[i] for i in range(n)) <= 6000, "vitamin_A_upper")
    
    # Vitamin C
    model.addConstr(gp.quicksum(vitamin_C_coeffs[i] * a[i] for i in range(n)) >= 45, "vitamin_C_lower")
    model.addConstr(gp.quicksum(vitamin_C_coeffs[i] * a[i] for i in range(n)) <= 1200, "vitamin_C_upper")
    
    # Calcium
    model.addConstr(gp.quicksum(calcium_coeffs[i] * a[i] for i in range(n)) >= 1300, "calcium_lower")
    model.addConstr(gp.quicksum(calcium_coeffs[i] * a[i] for i in range(n)) <= 3000, "calcium_upper")
    
    # Iron
    model.addConstr(gp.quicksum(iron_coeffs[i] * a[i] for i in range(n)) >= 8, "iron_lower")
    model.addConstr(gp.quicksum(iron_coeffs[i] * a[i] for i in range(n)) <= 40, "iron_upper")
    
    
    # Constraints for each ingredient to be less than 30% total calories
    total_calories = sum(calories_coeffs[i] * a[i] for i in range(n))
    for i in range(n):
        model.addConstr((calories_coeffs[i] * a[i]) <= 0.3 * total_calories, name=f"calories_item_{i}")
    
    # Constraints for each ingredient to be less than 30% total protein
    total_protein = sum(protein_coeffs[i] * a[i] for i in range(n))
    for i in range(n):
        model.addConstr((protein_coeffs[i] * a[i]) <= 0.3 * total_protein, name=f"protein_item_{i}")
    
    
    for i in range(n):
        model.addConstr(a[i] <= 3 , name=f"item_constraint{i}")
    
    # Optimize the model
    model.setParam('OutputFlag', 0)
    model.optimize()
    
    # Print the solution
    if model.status != GRB.OPTIMAL:
        continue
    else: 
        #for i in range(n):
            #print(f'{rand_food_lst[i]}: {a[i].x}')
        model_d_opt = model.objVal
        print(f'Seed: {z}')
        print(f'Minimum cost required: ${model.objVal:.2f}')
        model_d = results_df()
        model_d.index = model_d.index + 1
        print(model_d, '\n')


Seed: 1
Minimum cost required: $2.45
             Food Item  Servings        Serving Unit
1  Peppers, Sweet, Raw      2.50            1 Pepper
2      Roasted Chicken      0.59        1 lb chicken
3   Apple, Raw, w/Skin      1.83  1 Fruit,3/Lb,Wo/Rf
4               Grapes      3.00           10 Grapes
5               Bagels      2.04                1 Oz
6          Wheat Bread      1.28             1 Slice
7         Poached Eggs      0.27             Lrg Egg
8               Turkey      0.56                1 Oz 

Seed: 3
Minimum cost required: $2.45
             Food Item  Servings Serving Unit
1  Peppers, Sweet, Raw      2.50     1 Pepper
2                 Tofu      0.59    1/4 block
3   Spaghetti W/ Sauce      1.83    1 1/2 Cup
4               Banana      3.00      1 Fruit
5               Bagels      2.04         1 Oz
6          Wheat Bread      1.28      1 Slice
7       Scrambled Eggs      0.27        1 Egg
8                 Beef      0.56         1 Oz 

Seed: 7
Minimum cost required: 

### Recommendation and Conclusion:

Combining these 6 lists with the optimal solution from part C, we now have 7 lists of food items with variety for the one-week meal plan.

Here is our recommendation for a balanced and diverse 7-day meal plan:
 
Monday:
Cost per child ($2.60) comprised of the following food items:
- 0.13 Servings of Carrots, Raw
- 0.68 Servings of Lettuce, Iceberg, Raw
- 2.92 Servings of Potatoes, Baked
- 0.55 Servings of Spaghetti W/ Sauce
- 2.91 Servings of Apple, Raw, w/Skin
- 0.82 Servings of Oranges
- 2.04 Servings of 2% Lowfat Milk
- 1.57 Servings of Skim Milk
- 0.25 Servings of White Rice
- 0.56 Servings of Pork
 
Tuesday:
Cost per child ($2.45) comprised of the following food items:
- 2.50 Servings of Peppers, Sweet, Raw
- 0.59 Servings of Roasted Chicken
- 1.83 Servings of Apple, Raw, w/Skin
- 3.00 Servings of Grapes
- 2.04 Servings of Bagels
- 1.28 Servings of Wheat Bread
- 0.27 Servings of Poached Eggs
- 0.56 Servings of Turkey
 
Wednesday:
Cost per child ($2.45) comprised of the following food items:
- 2.50 Servings of Peppers, Sweet, Raw
- 0.59 Servings of Tofu
- 1.83 Servings of Spaghetti W/ Sauce
- 3.00 Servings of Banana
- 2.04 Servings of Bagels
- 1.28 Servings of Wheat Bread
- 0.27 Servings of Scrambled Eggs
- 0.56 Servings of Beef
 
Thursday:
Cost per child ($2.50) comprised of the following food items:
- 3.00 Servings of Potatoes, Baked
- 0.59 Servings of Spaghetti W/ Sauce
- 1.20 Servings of Tomato, Red, Ripe, Raw
- 3.00 Servings of Grapes
- 2.04 Servings of Oranges
- 1.25 Servings of Bagels
- 0.22 Servings of Poached Eggs
- 0.59 Servings of Turkey

Friday:
Cost per child ($2.68) comprised of the following food items:
- 0.08 Servings of Broccoli
- 0.10 Servings of Carrots, Raw
- 2.33 Servings of Peppers, Sweet, Raw
- 0.54 Servings of Roasted Chicken
- 3.00 Servings of Tomato, Red, Ripe, Raw
- 1.57 Servings of Apple, Raw, w/Skin
- 2.04 Servings of Wheat Bread
- 1.69 Servings of White Bread
- 0.56 Servings of Turkey

Saturday:
Cost per child ($2.80) comprised of the following food items:
- 0.03 Servings of Potatoes, Baked
- 2.03 Servings of Roasted Chicken
- 3.00 Servings of Spaghetti W/ Sauce
- 3.00 Servings of Apple, Raw, w/Skin
- 2.52 Servings of Grapes
- 3.00 Servings of Kiwifruit, Raw, Fresh
- 2.04 Servings of Oranges
- 1.18 Servings of Bagels
- 0.09 Servings of Skim Milk
- 0.15 Servings of Scrambled Eggs
- 0.56 Servings of Beef 

Sunday:
Cost per child ($2.45) comprised of the following food items:
- 2.50 Servings of Peppers, Sweet, Raw
- 0.59 Servings of Tofu
- 1.83 Servings of Spaghetti W/ Sauce
- 3.00 Servings of Banana
- 2.04 Servings of Oranges
- 1.28 Servings of Bagels
- 0.27 Servings of Scrambled Eggs
- 0.56 Servings of Turkey


Our recommended one-week meal plan ensured keeping the daily cost per kid at the minimum, with each day’s cost not exceeding $3 per kid. 

As we conclude our analysis, we wish to convey our hope for the future growth and improvement of the “Feeding Hope” initiative project to continue their focus on potential opportunities in minimizing cost, as well as improving nutrient intake and enhancing the variety of food consumed by children. Since the town of Starlight is still in the process of recovery during the time of this analysis, the resources in the town are limited. Hence, our analysis focused mainly on minimizing costs in providing nutritious meals. 

However, while the town’s economic conditions improve, there is potential for the meals to include an even wider variety of different food group items, ensuring children can receive higher quality meals to encourage their growth and strengthen their minds and bodies. With gradual improvement in the economic outlook of Starlight, we hope that the “Feeding Hope” project efforts will remain successful in supporting the town to provide nutritious meals for the children, as it progresses into a better future.

<b>Reference:</b>

[1]https://cichprofile.ca/module/9/section/2/page/recommended-number-of-servings-per-day-for-canadian-children-and-youth-aged-2-to-13-years-by-age-group-canada-2011/ 

